docs/adr/0006-json-schemas-for-messaging.md

```diff
@@ -0,0 +1,98 @@
+
+## Decision Outcome
+
+PLACEHOLDER
+
+### [option 1] Keep existing approach
+
+Currently our documentation drifts if a message format is tweaked and the docs are not updated. Ideally
+our code review process would catch this but sometimes code subleties are not caught that alter the
+message format. Validation changes can also fail to appear in the documentation, even though what is
+considered a valid message key/value should be documented.
+
+We have to occasionally duplicate `joi` schemas between repos, which should likely be moved to `fxa-shared`
+to benefit re-use, but this does not help non-Node projects consume our messages.
+
+- Pros
+  - No additional work is needed that we don't already do.
+- Cons
+  - Our documentation occasionally drifts.
+  - It's additional work to maintain accurate `joi` schemas spread throughout the code-base _or_
+  - It's additional work to consolidate our `joi` schemas in `fxa-shared`.
+  - `joi` schemas can't be used by consumers not written in JS.
+
+### [option 2] Use protobuf
+
+Using protobuf would allow us to validate and document valid message formats that would also consume
+less space and be more performant to serialize/deserialize. Validation is built-in as invalid messages
+will not deserialize, and documenation can be generated based on the protobuf spec for a message.
+
+Protobuf tooling is widely available for most langauges.
+
+- Pros
+  - High throughput messaging could benefit from higher efficiency of serializing/deserializing of messages
+  - Documentation stays in sync with message spec
+  - Validation is built-in to the message serialization/deserialization process
+- Cons
+  - Messages are in binary and not easily introspectable
+  - Existing message consumers will need to be updated to handle protobuf
+  - Migration procedure will need to be created/implemented to shift from existin JSON format
+
+### [option 3] Use JSON-Schemas
+
+JSON-Schemas can define existing FxA JSON messages used in our queue system. There are libraries in
+most languages to consume these schemas for validation purposes, and libraries exist to generate
+documenation from a JSON-Schema.
+
+[SchemaVer] can be introduced to existing FxA JSON messages as consumers ignore unknown keys, while
+consumers using the JSON-Schema can verify they can handle new message formats or if they're incompatible.
+Some of this logic is manual, as [SchemaVer] is not an official part of the JSON-Schema specification. There
+is a recommendation of [SchemaVer] in [Best Web Practices for Data](https://www.w3.org/TR/dwbp/#dataVersioning).
+
+- Pros
+  - No changes needed for existing consumers
+  - Validation can be done based on the JSON-Schema without custom code
+  - Documentation can be generated to stay in sync with message spec
+  - JSON is human readable without further processing.
+  - JSON is widely used on the web and our increasing use of JWTs also result in JSON.
+  - JSON-Schema is widely supported in a variety of langauges.
```

Comment:
Are there good libraries to consume JSON schema in both Hapi and Express? Without support, migrating to JSON Schema might just result in "yet another set of docs that are out of date". I see https://www.npmjs.com/package/express-jsonschema and https://github.com/simon-p-r/hapi-json-schema, https://github.com/bloglovin/timplan. I have no idea of their quality and at least the Hapi libraries do not seem to have been updated in quite a while.

Reply:
This ADR is intended strictly for message formats, not API payloads (though perhaps we should represent the JWT payloads with JSON-Schema). For API's I think we should use gRPC and/or Swagger, which do this kind of stuff.

Reply:
OK, in my mind I broadened the scope of where this applies.

Reply:
I'll prolly come up with a Swagger/gRPC ADR next. 😉

Reply:
> Are there good libraries to consume JSON schema in both Hapi and Express?

I'm not sure about hapi or express connectors, but AJV is a JSON Schema validator that's widely used.

Moving from a JS-based schema (expressed in Joi) to a JSON Schema will reduce what we can do (JSON Schema isn't a full-fledged programming language), but I'm not sure how complex our Joi validation currently is. There are a couple of libraries that convert back and forth between Joi and JSON Schema, not sure about their quality, but if there's interest, we could play with auto-converting some of the stuff, then manually extracting out duplicated types using `$id` and `$ref`.

See also the list of tools here: http://json-schema.org/implementations.html

> For API's I think we should use gRPC and/or Swagger, which do this kind of stuff.

Swagger is a really good option to consider; I think the tooling is better, but I'm not sure how the tooling we'd actually use compares, and I'm not sure what else we'd get vs JSON (Hyper-)Schema. I imagine the security folks, and maybe QA, would be able to use more off-the-shelf testing tools if we opted for Swagger. Given that the FxA APIs are RESTful, gRPC and its protobuf transport seems like less of a clear fit. All that said, I think it'd be good to dig into the use cases carefully (in a separate ADR).